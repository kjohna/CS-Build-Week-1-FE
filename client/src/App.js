import React from "react";
import { Switch, Route, Redirect } from "react-router-dom";
import styled from "styled-components";

import { useAuthContext } from "./AuthProvider";
import Adv from "./Adv";
import Login from "./Login";

const AppContainer = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  height: 100vh;
`;

const PrivateRoute = ({ component, ...options }) => {
  const { user } = useAuthContext();
  console.log(user);
  const finalComponent = user ? component : Login;

  return <Route {...options} component={finalComponent} />;
};

function App() {
  // protect routes from user w/o token
  let routes = (
    <Switch>
      <Route path="/" exact component={Login} />
      <PrivateRoute path="/adv" component={Adv} />
    </Switch>
  );

  return <AppContainer>{routes}</AppContainer>;
}

export default App;
