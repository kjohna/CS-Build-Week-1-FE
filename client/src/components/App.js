import React from "react";
import { Switch, Redirect, Route, useHistory } from "react-router-dom";
import { connect } from "react-redux";
import styled from "styled-components";

import Adv from "./Adv";
import LoginRegister from "./LoginRegister";

import actionExports from "../store/actions";

const { advInit } = actionExports;

const AppContainer = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  height: 100vh;
`;

// privateRoute middleware, protects a route based on login state (token present in redux store)
const PrivateRoute = ({ component, token, ...options }) => {
  const finalComponent = token ? (
    <Route {...options} component={component} />
  ) : (
    <Redirect to="/" />
  );

  return finalComponent;
};

function App(props) {
  // protect routes from user w/o token
  const { auth, advInit } = props;
  const token = auth.token;
  const history = useHistory();
  // if no token, check local storage for existing token
  // auth.error is set after unsuccessful lcl storage check
  if (!token && !auth.error && !auth.loading) {
    // will redirect to /adv if token is found
    // need to pass in history and location to achieve this since token is not set, a redirect to "/" will render this time
    // NOTE: could employ useLocation to do this dynamically
    advInit("/adv", history);
  }
  // advInit(props.dispatch);
  let routes = (
    <Switch>
      <Route exact path="/" component={LoginRegister} />
      <PrivateRoute path="/adv" component={Adv} token={token} />
    </Switch>
  );

  return <AppContainer>{routes}</AppContainer>;
}

const mapStateToProps = state => ({
  auth: state.auth
});

const mapDispatchToProps = {
  advInit
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(App);
