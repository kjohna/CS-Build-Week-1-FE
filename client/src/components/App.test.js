import React from "react";
import ReactDOM from "react-dom";
import App from "./App";
import reducer from "../store/reducers";
// need enzyme since we're using redux
// https://circleci.com/blog/continuously-testing-react-applications-with-jest-and-enzyme/
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
// this could go in a setup file
// https://airbnb.io/enzyme/docs/installation/index.html
import { configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

configure({ adapter: new Adapter() });
// end setup

it("renders without crashing", () => {
  const emptyState = reducer();
  const props = {
    dispatch: jest.fn(),
    store: {
      getState: () => emptyState
    }
  };
  const wrapper = shallow(<App {...props} />);
  expect(toJson(wrapper)).toMatchSnapshot();
});
