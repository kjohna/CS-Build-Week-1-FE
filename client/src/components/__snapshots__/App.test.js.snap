// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<ContextProvider
  value={null}
>
  <App
    advInit={[Function]}
    auth={
      Object {
        "error": null,
        "loading": false,
        "token": "",
        "username": "",
      }
    }
    dispatch={[MockFunction]}
    store={
      Object {
        "getState": [Function],
      }
    }
  />
</ContextProvider>
`;
